src/libServer/LookupServer.cpp

```diff
@@ -1423,12 +1430,115 @@ Json::Value LookupServer::GetEthSyncing() {
   return Json::Value(false);
 }
 
-Json::Value LookupServer::GetEthAccounts() {
+Json::Value LookupServer::GetEmptyResponse() {
   LOG_MARKER();
   const Json::Value expectedResponse = Json::arrayValue;
   return expectedResponse;
 }
 
+Json::Value LookupServer::GetEthStorageAt(std::string const& address,
+                                          std::string const& position,
+                                          std::string const& /*blockNum*/) {
+  LOG_MARKER();
+
+  string vname{};
+  Json::Value indices = Json::arrayValue;
+
+  if (Mediator::m_disableGetSmartContractState) {
+    LOG_GENERAL(WARNING, "API disabled");
+    throw JsonRpcException(RPC_INVALID_REQUEST, "API disabled");
+  }
+
+  if (!LOOKUP_NODE_MODE) {
+    throw JsonRpcException(RPC_INVALID_REQUEST, "Sent to a non-lookup");
+  }
+
+  try {
+    Address addr{ToBase16AddrHelper(address)};
+    shared_lock<shared_timed_mutex> lock(
+        AccountStore::GetInstance().GetPrimaryMutex());
+
+    const Account* account = AccountStore::GetInstance().GetAccount(addr, true);
+
+    if (account == nullptr) {
+      throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
+                             "Address does not exist");
+    }
+
+    if (!account->isContract()) {
+      throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
+                             "Address not contract address");
+    }
+    LOG_GENERAL(INFO, "Contract address: " << address);
+    Json::Value root;
+    const auto indices_vector =
+        JSONConversion::convertJsonArrayToVector(indices);
+
+    if (!account->FetchStateJson(root, vname, indices_vector)) {
+      throw JsonRpcException(RPC_INTERNAL_ERROR, "FetchStateJson failed");
+    }
+
+    // Attempt to get storage at position.
+    // Left-pad position with 0s up to 64
+    std::string zeroes =
+        "0000000000000000000000000000000000000000000000000000000000000000";
+
+    if (position.size() > zeroes.size()) {
+      throw JsonRpcException(RPC_INTERNAL_ERROR,
+                             "Position string is too long! " + position);
+    }
+
+    if (!position.empty()) {
```

Comment:
I don't think I fully understand the algorithm here, but that's probably not something that does leftPad. Also please note you dereference `end()` iterator in line 1497 on the first which is basically UB.
It could be a nice candidate for STL algorithms. Please consider the following:
`const auto firstPosOfX = position.find_first_of('x');
if(firstPosOfX == std::string::npos) {
   // Handle errors accordingly
}
// Calculate valid range
const auto maxEnd = std::min(zeros.size() - firstPosOfX, positions.size() - firstPosOfX);
// Copy positions to zeroed pad
std::copy(positions.begin() + firstPosOfX, positions.begin() + firstPosOfX + maxEnd, zeros.begin() + firstPosOfX);
`
Obviously this requires some attention and testing.

Reply:
👍

Reply:
I think I made it slightly cleaner/safer using replace (on other branch - check it out?)

Reply:
Yeah I checked there, looks ok, thanks.